Support for the verification with class hash

marijamijailovic opened the pull request:
Closes #2668 

## Introduced changes

This PR adds support for `sncast verify --verifier walnut` to use class hash.

## Checklist

<!-- Make sure all of these are complete -->

- [ ] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

marijamijailovic committed:
Support for the verification with class hash

kkawula commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -113,9 +123,13 @@ impl VerificationInterface for WalnutVerificationInterface {
 #[command(about = "Verify a contract through a block explorer")]
 pub struct Verify {
+    /// Class hash of a contract to be verified
+    #[clap(short = 'x', long)]
```
```suggestion
    #[clap(short = 'g', long)]
```

kkawula commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -67,10 +72,15 @@ impl VerificationInterface for WalnutVerificationInterface {
         // Serialize the JSON object to a JSON string
         let source_code = serde_json::Value::Object(file_data);
 
+        // Convert contract addresss and class hash from the Felt to String
```
```suggestion
        // Convert contract address and class hash from the Felt to String
```

kkawula reviewed (commented):
Please add tests for new logic

marijamijailovic committed:
Added test for --class-hash, shourtcode changed -x to -g, fix typo

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Removed verify contract section from sncast/README.md

marijamijailovic commented:
Hi @kkawula, thanks for your comments! I have addressed them. Additionally, I have removed the "Verify contract" section from sncast/README.md due to a failed test (changing the --network parameter to --url "http://127.0.0.1:5055"), as Walnut does not support the --url parameter.

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

kkawula reviewed (approved):
Please add an entry in `CHANGELOG.md`

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Added changes to the changelog

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic commented:
Hey @kkawula, I’ve updated my changes with the latest changes from the master branch. Let me know if there’s anything else needed.

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Merge with master and resolve conflicts

marijamijailovic committed:
Added ArgGroup, and ContractIdentifier enum

marijamijailovic commented on docs/src/starknet/verify.md:
```diff
@@ -45,7 +45,7 @@ $ sncast \
 
 command: verify
-message: Contract verification has started. You can check the verification status at the following link: https://api.walnut.dev/v1/verification/77f1d905-fdb4-4280-b7d6-57cd029d1259/status.
+message: Contract verification has started. You can check the verification status at the following link: https://app.walnut.dev/verification/status/77f1d905-fdb4-4280-b7d6-57cd029d1259.
```
![CleanShot 2025-03-03 at 12  11 11](https://github.com/user-attachments/assets/a7279891-2fbe-42d9-9e23-3ea20723bcd5)
Hey, when you open the link, you will see the data in the photo I posted. You can use our search for class_hash and check out the source code.

cptartur replied:
What I meant that in the command above you run `sncast verify --class-hash 0x0227f52a4d2138816edf8231980d5f9e6e0c8a3deab45b601a1fcee3d4427b02` but the link leads to verification result of class hash `0x031966c9fe618bcee61d267750b9d46e3d71469e571e331f35f0ca26efe306dc`.

This is not critical however.

marijamijailovic replied:
Thanks for pointing this out. I have updated the docs, so now the class hash and verification ID are connected.

marijamijailovic commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -113,9 +123,13 @@ impl VerificationInterface for WalnutVerificationInterface {
     /// Address of a contract to be verified
     #[clap(short = 'd', long)]
-    pub contract_address: Felt,
+    pub contract_address: Option<Felt>,
```
Thanks, I have added ArgGroup!

marijamijailovic commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -151,24 +165,33 @@ impl fmt::Display for Verifier {
+        Ok(())
+    }
+}
+
```
I have removed this one!

marijamijailovic commented on crates/sncast/src/starknet_commands/verify.rs:
```diff
@@ -38,7 +42,8 @@ impl VerificationInterface for WalnutVerificationInterface {
         &self,
-        contract_address: Felt,
+        class_hash: Option<Felt>,
+        contract_address: Option<Felt>,
```
Thanks, added enum!

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic commented:
Hi @cptartur , thanks for your comments! I have addressed them and update my branch so it's up to date with master.

cptartur commented on CHANGELOG.md:
There are duplicate entries in the file

cptartur commented on CHANGELOG.md:
```diff
@@ -28,6 +34,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `snforge_scarb_plugin` now emits an error when parameters are passed without using the `#[fuzzer]` attribute
 - A bug that was causing execution to hang if using forking
 
+### Cast
+
+#### Changed
+
+- `verify` command now supports the `--class-hash` for Walnut verification
+
```
```suggestion
```

cptartur commented on crates/sncast/src/starknet_commands/verify/mod.rs:
```diff
@@ -85,10 +94,26 @@ pub async fn verify(
+        },
+
+        _ => {
+            return Err(anyhow!(
+                "Exactly one of class_hash or contract_address must be provided."
```
This is not a recoverable error, clap logic for args verification should now prevent this arm from ever being reached.

```suggestion
            unreachable!(
                "Exactly one of class_hash or contract_address must be provided."
```

marijamijailovic replied:
Fixed

marijamijailovic committed:
Fix changelog double entry, update verify.md, and add unreachable!

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Update changelog.md

marijamijailovic committed:
Revert formatting changes in changelog.md

cptartur commented on CHANGELOG.md:
```diff
@@ -1092,4 +1096,4 @@ snforge_std = { git = "https://github.com/foundry-rs/starknet-foundry", tag = "v
 #### Added
 
-- Initial release
+- Initial release
```
Should this be modified?

marijamijailovic replied:
Removed this empty line

cptartur reviewed (approved):
Please address the changelong change, otherwise it's good

cptartur commented:
@ddoktorski can you take look at this as well?

cptartur commented:
@kkawula please re-review.

marijamijailovic committed:
Remove empty line diff in changelog.md

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic commented:
Thank you for all your previous comments! I've updated the branch based on the feedback. Could you please take another look and let me know if it's ready for merging? Appreciate your time!

cptartur commented:
> Thank you for all your previous comments! I've updated the branch based on the feedback. Could you please take another look and let me know if it's ready for merging? Appreciate your time!

@marijamijailovic good on my side, let's wait for @kkawula review

kkawula commented on crates/sncast/README.md:
```diff
@@ -225,4 +225,4 @@ $HOME/.asdf/shims/scarb
 <br>
 
-If you previously installed scarb using official installer, you may need to remove this installation or modify your PATH to make sure asdf installed one is always used.
+If you previously installed scarb using official installer, you may need to remove this installation or modify your PATH to make sure asdf installed one is always used.
```
Please revert this change

marijamijailovic replied:
Reverted

kkawula commented on docs/src/appendix/sncast/verify.md:
```diff
@@ -1,8 +1,15 @@
-Required.
+## `--class-hash, -g <CLASS_HASH>`
+
+Required if `--contract-address` is not provided.
```
```suggestion
Optional. Required if `--contract-address` is not provided.
```

marijamijailovic replied:
Added

kkawula commented on docs/src/appendix/sncast/verify.md:
```diff
@@ -1,8 +1,15 @@
+
+## `--contract-address, -d <CONTRACT_ADDRESS>`
+
+Required if `--class-hash` is not provided.
```
```suggestion
Optional. Required if `--class-hash` is not provided.
```

marijamijailovic replied:
Added

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Revert sncast/readme changes, add optional to the verify.md

marijamijailovic commented:
Hey @cptartur, I see that this PR has been removed from the merge queue. Should I resolve the merge conflict?

cptartur commented:
> Hey @cptartur, I see that this PR has been removed from the merge queue. Should I resolve the merge conflict?

@marijamijailovic yes, please.

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic commented:
> > Hey @cptartur, I see that this PR has been removed from the merge queue. Should I resolve the merge conflict?
> 
> @marijamijailovic yes, please.

I have updated branch

cptartur commented on CHANGELOG.md:
```diff
@@ -36,6 +36,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 #### Changed
 
+- `verify` command now supports the `--class-hash` for Walnut verification
+
```
Please move this to `### Unreleased` section and we can merge.

marijamijailovic replied:
Moved

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic committed:
Move verify to unrelease section

cptartur commented:
@marijamijailovic I think you need to resolve conflicts again and it will start merging

marijamijailovic committed:
Merge remote-tracking branch 'upstream/master' into verification-with-class-hash

marijamijailovic commented:
> @marijamijailovic I think you need to resolve conflicts again and it will start merging

Hey, @cptartur I resolved the conflicts.
